In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

In [155]:
import general_utils as Ugen
import matplotlib.pyplot as plt
import pandas as pd
import math
from pandas.tools.plotting import scatter_matrix
import numpy as np
import glob, os
import requests

In [3]:
##GET DATA
#GET FILES
main_dir=os.getcwd()
allFiles = glob.glob(main_dir.split('/Studies')[0]+'/toronto_parking_data'+ "/*.csv")

#LOAD TO DATAFRAME
list_ = []
for file_path in allFiles:
    try:
        df = pd.read_csv(file_path,index_col=None,quoting=2,error_bad_lines=False) #quoting=2
        list_.append(df)
        print ('successfully loaded: '+ file_path)
    except Exception as err:
        if 'EOF following escape character' in str(err):
            try:
                df = pd.read_csv(file_path,index_col=None,encoding='utf-16',error_bad_lines=False) #quoting=2
                list_.append(df)
                print ('successfully loaded: '+ file_path)
            except Exception as err:
                print ('NOT loaded: '+ file_path)
                print (str(err))
        else:
            print ('NOT loaded: '+ file_path)
            print (str(err))
tpt_df = pd.concat(list_)
print (tpt_df.shape)
tpt_df.head()

b'Skipping line 181770: expected 11 fields, saw 12\n'
b'Skipping line 170864: expected 11 fields, saw 12\n'

successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_data_2008.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_data_2009.csv


b'Skipping line 1075521: expected 11 fields, saw 14\n'
b'Skipping line 1413505: expected 11 fields, saw 16\n'


successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_data_2010.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_data_2011.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2012.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2013.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2014_1.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2014_2.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2014_3.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/Parking_Tags_Data_2014_4.csv
successfully loaded: /Users/whitesi/Documents/Programming/Python/toronto_parking_data/P

tag_number_masked  date_of_infraction  infraction_code  \
0          ***63611            20080101                3   
1          ***99026            20080101                9   
2          ***63613            20080101                3   
3          ***63617            20080101                3   
4          ***63608            20080101                3   

           infraction_description  set_fine_amount  time_of_infraction  \
0  PARK/LEAVE ON PRIVATE PROPERTY               30                 NaN   
1    STOP HWY PROHIBITED TIME/DAY               60                 NaN   
2  PARK/LEAVE ON PRIVATE PROPERTY               30                 NaN   
3  PARK/LEAVE ON PRIVATE PROPERTY               30                 NaN   
4  PARK/LEAVE ON PRIVATE PROPERTY               30                 NaN   

  location1         location2 location3 location4 province  
0       NaN     364 EAST MALL       NaN       NaN       ON  
1       NaN  80 RICHMOND ST W       NaN       NaN       ON  
2       NaN     364 EAST MALL       NaN       NaN       ON  
3       NaN     362 EAST MALL       NaN       NaN       ON  
4       NaN     366 EAST MALL       NaN       NaN       ON

In [8]:
tpt_df=tpt_df[pd.notnull(tpt_df['location2'])]
tpt_df=tpt_df.drop(['tag_number_masked','location3','location4','province'],axis=1)
print (tpt_df.shape)

##19 million parking ticket records.....lol


(19412524, 7)


In [29]:
locations=tpt_df['location2'].unique()
len(locations)

568146

In [10]:
len(locations)/2500 ##number of days it will take to get the data with free google api (2500 req/day)

227.2584

So obviously we can't get lat/lon for all locations...let's do it for the top X grossing spots...

In [22]:
tpt_df['year']=tpt_df['date_of_infraction'].map(lambda x: int(str(x)[0:4])) #
tpt_df.head()

date_of_infraction  infraction_code          infraction_description  \
0            20080101                3  PARK/LEAVE ON PRIVATE PROPERTY   
1            20080101                9    STOP HWY PROHIBITED TIME/DAY   
2            20080101                3  PARK/LEAVE ON PRIVATE PROPERTY   
3            20080101                3  PARK/LEAVE ON PRIVATE PROPERTY   
4            20080101                3  PARK/LEAVE ON PRIVATE PROPERTY   

   set_fine_amount  time_of_infraction location1         location2  year  
0               30                 NaN       NaN     364 EAST MALL  2008  
1               60                 NaN       NaN  80 RICHMOND ST W  2008  
2               30                 NaN       NaN     364 EAST MALL  2008  
3               30                 NaN       NaN     362 EAST MALL  2008  
4               30                 NaN       NaN     366 EAST MALL  2008

In [42]:
infraction_ids=list(tpt_df.infraction_code.unique())

In [43]:
tpt_grouped_df=pd.DataFrame({'count' : tpt_df.groupby( ['year','infraction_code','location2'] ).size()}).reset_index()

In [44]:
print(tpt_grouped_df.shape)
tpt_grouped_df.head()
# tpt_grouped_df[tpt_grouped_df['year']==2015]

(2833240, 4)


year  infraction_code           location2  count
0  2008                1    1  FOUR WINDS DR      1
1  2008                1     1 9 MOBERLY AVE      1
2  2008                1         1 BANGOR RD      2
3  2008                1         1 BEAMAN RD      2
4  2008                1  1 BEDFORD PARK AVE      8

In [47]:
##now calculate total amount...
##create dict of infraction_code:fine_amount
fine_amount_dict={ic:{'fine_amt':tpt_df[tpt_df['infraction_code']==ic].set_fine_amount.median(),
                      'fine_descp':tpt_df[tpt_df['infraction_code']==ic].infraction_description.iloc[0]}    
                          for ic in infraction_ids}

def calc_fine_sum(df):
    return df['count']*fine_amount_dict[df['infraction_code']]['fine_amt']

def get_fine_descrp(df):
    return fine_amount_dict[df['infraction_code']]['fine_descp']


tpt_grouped_df['fine_sum']=tpt_grouped_df.apply(calc_fine_sum,axis=1)
tpt_grouped_df['fine_descrp']=tpt_grouped_df.apply(get_fine_descrp,axis=1)
tpt_grouped_df.sort_values('fine_sum',ascending=False)

year  infraction_code               location2  count  fine_sum  \
2779875  2015              347       1090 DON MILLS RD   1889    472250   
2426803  2014              367          410 COLLEGE ST    742    333900   
1883194  2012              367          410 COLLEGE ST    672    302400   
818428   2010                3        1750 FINCH AVE E   9574    287220   
2784954  2015              367          410 COLLEGE ST    604    271800   
450759   2009                3        1750 FINCH AVE E   8669    260070   
2421668  2014              347       1090 DON MILLS RD   1028    257000   
53976    2008                3        25 THE WEST MALL   8291    248730   
2408008  2014              257          1 BRIMLEY RD S   2296    241080   
2768535  2015              257          1 BRIMLEY RD S   2267    238035   
1193779  2011                3        1750 FINCH AVE E   7867    236010   
1568669  2012                3        1750 FINCH AVE E   7615    228450   
50294    2008                3        1750 FINCH AVE E   7525    225750   
820155   2010                3        2075 BAYVIEW AVE   7509    225270   
1521399  2011              355        555 REXDALE BLVD    469    211050   
1518927  2011              347        3401 DUFFERIN ST    830    207500   
1523025  2011              367          410 COLLEGE ST    460    207000   
454669   2009                3        25 THE WEST MALL   6698    200940   
2781648  2015              347         45 OVERLEA BLVD    787    196750   
1195436  2011                3        2075 BAYVIEW AVE   6482    194460   
855235   2010                5            20 EDWARD ST   4846    193840   
452644   2009                3        2075 BAYVIEW AVE   6312    189360   
1230197  2011                5            20 EDWARD ST   4633    185320   
1570277  2012                3        2075 BAYVIEW AVE   5931    177930   
488695   2009                5            20 EDWARD ST   4400    176000   
90392    2008                5            20 EDWARD ST   4328    173120   
2423455  2014              347         45 OVERLEA BLVD    691    172750   
2139580  2014                3        2075 BAYVIEW AVE   5521    165630   
1603694  2012                5            20 EDWARD ST   4132    165280   
2170330  2014                5            20 EDWARD ST   4093    163720   
...       ...              ...                     ...    ...       ...   
403486   2008              348       405 SHERBOURNE ST      1         0   
403485   2008              348       404 OLD WESTON RD      1         0   
403484   2008              348      4012 SHEPPARD AV E      1         0   
403483   2008              348     4002 SHEPPARD AVE E      1         0   
403482   2008              348      4002 SHEPPARD AV E      1         0   
403502   2008              348        4438 SHEPPARD AV      9         0   
403504   2008              348       4438 SHEPPARD AVE      3         0   
403526   2008              348         5 SHADY GOLFWAY      1         0   
403516   2008              348  47 THORNCLIFFE PARK DR      2         0   
403525   2008              348         5 NORTHTOWN WAY      5         0   
403524   2008              348           5 FARNHAM AVE      1         0   
403523   2008              348          5 ELLESMERE RD      1         0   
403522   2008              348            5 BLEVINS PL      1         0   
403521   2008              348        4913 BATHURST ST      1         0   
403520   2008              348          4800 LESLIE ST      2         0   
403519   2008              348        480 DANFORTH AVE      3         0   
403518   2008              348    475 COMMISSIONERS ST      1         0   
403517   2008              348           47 WYATT WALK      1         0   
403515   2008              348           4683 YONGE ST      1         0   
403505   2008              348        4455 BATHURST ST      1         0   
403514   2008              348        464 DANFORTH AVE      1         0   
403513   2008              348    

In [56]:
fine_amount_dict[347]

{'fine_amt': 250.0, 'fine_descp': 'PARK IN A FIRE ROUTE'}

This is wild...so in 2015 there were 1889 parking tickets issued at one location (1090 Don Mills Rd).

That's just over 5 tickets a day on average...and that's if you assume every day at the mall/this location has the same traffic..which is probably not true. In reality, this would be concentrated on weekends so it could be something like 10-20 tickets (random guess) on weekends.

If you assume each ticketed vehicle is there for at least an hour...that's a good chunk of the day where the fire route is at least partially blocked..obviously this needs to be treated more as an actual safety concern and less as an incredible revenue generating source for Toronto's municipal government. 

Obviously this rule is not clear...I don't think people would risk a $250 ticket just to get a closer parking spot..this spot (and likely many others) clearly needs better labelling. Putting a sign on the wall is one thing...how about the government takes $20 from the $472,250 they earned in 2015 alone (okay, maybe 50 percent of the people fought their tickets or didn't pay them...only $236,125) and puts some red paint on the pavement. 




In [58]:
[{infrac_type:infrac_dict['fine_descp']} for infrac_type,infrac_dict in fine_amount_dict.items() if infrac_dict['fine_amt']==0]

[{130: 'STOP PROH TIME MUTUAL STREET'},
 {150: 'PARK VEH BEYOND DESIG AREA'},
 {166: 'PARK ON TRINITY DRIVE'},
 {197: 'PARK NOT WITHIN PERMITTED TIME'},
 {251: 'LEAVE IN PARK NOT DESIG AREA'},
 {252: 'PARK IN PARK - PROHIBITED AREA'},
 {256: 'LEAVE OVERNIGHT WITHOUT PERMIT'},
 {258: 'STOP CONTRARY POSTED CONDITION'},
 {348: 'PARK IN DISABLED NO PERMIT'}]

In [55]:
tpt_df[(tpt_df['infraction_code']==348) & (tpt_df['year']==2008)]

date_of_infraction  infraction_code      infraction_description  \
541819             20080314              348  PARK IN DISABLED NO PERMIT   
541827             20080314              348  PARK IN DISABLED NO PERMIT   
541850             20080314              348  PARK IN DISABLED NO PERMIT   
543826             20080314              348  PARK IN DISABLED NO PERMIT   
544495             20080314              348  PARK IN DISABLED NO PERMIT   
545496             20080314              348  PARK IN DISABLED NO PERMIT   
546027             20080314              348  PARK IN DISABLED NO PERMIT   
546321             20080314              348  PARK IN DISABLED NO PERMIT   
547188             20080314              348  PARK IN DISABLED NO PERMIT   
548703             20080314              348  PARK IN DISABLED NO PERMIT   
549431             20080314              348  PARK IN DISABLED NO PERMIT   
549568             20080314              348  PARK IN DISABLED NO PERMIT   
549750             20080314              348  PARK IN DISABLED NO PERMIT   
549772             20080314              348  PARK IN DISABLED NO PERMIT   
549788             20080314              348  PARK IN DISABLED NO PERMIT   
549839             20080314              348  PARK IN DISABLED NO PERMIT   
549875             20080314              348  PARK IN DISABLED NO PERMIT   
549970             20080314              348  PARK IN DISABLED NO PERMIT   
550138             20080314              348  PARK IN DISABLED NO PERMIT   
550315             20080314              348  PARK IN DISABLED NO PERMIT   
550529             20080314              348  PARK IN DISABLED NO PERMIT   
550710             20080315              348  PARK IN DISABLED NO PERMIT   
550978             20080315              348  PARK IN DISABLED NO PERMIT   
553682             20080315              348  PARK IN DISABLED NO PERMIT   
554186             20080315              348  PARK IN DISABLED NO PERMIT   
554363             20080315              348  PARK IN DISABLED NO PERMIT   
554461             20080315              348  PARK IN DISABLED NO PERMIT   
554468             20080315              348  PARK IN DISABLED NO PERMIT   
555019             20080315              348  PARK IN DISABLED NO PERMIT   
555312             20080315              348  PARK IN DISABLED NO PERMIT   
...                     ...              ...                         ...   
2734482            20081210              348  PARK IN DISABLED NO PERMIT   
2734520            20081210              348  PARK IN DISABLED NO PERMIT   
2734799            20081210              348  PARK IN DISABLED NO PERMIT   
2735368            20081210              348  PARK IN DISABLED NO PERMIT   
2738330            20081210              348  PARK IN DISABLED NO PERMIT   
2739678            20081210              348  PARK IN DISABLED NO PERMIT   
2741889            20081211              348  PARK IN DISABLED NO PERMIT   
2752165            20081212              348  PARK IN DISABLED NO PERMIT   
2753857            20081212              348  PARK IN DISABLED NO PERMIT   
2755336            20081212              348  PARK IN DISABLED NO PERMIT   
2763364            20081213              348  PARK IN DISABLED NO PERMIT   
2763579            20081213              348  PARK IN DISABLED NO PERMIT   
2775898            20081215              348  PARK IN DISABLED NO PERMIT   
2777583            20081215              348  PARK IN DISABLED NO PERMIT   
2786335            20081216              348  PARK IN DISABLED NO PERMIT   
2790086            20081217              348  PARK IN DISABLED NO PERMIT   
2813972            20081221              348  PARK IN DISABLED NO PERMIT   
2816192            20081222              348  PARK IN DISABLED NO PERMIT   
2818430            20081222              348  PARK IN DISABLED NO PERMIT   
2820395            20081223              348  PARK IN DISABLED NO PERMIT   
2825255            20081223              348  PARK IN DIS

Appears that they stopped fining people for parking without a permit?? or perhaps this is a data issue??

In [60]:
tpt_df[(tpt_df['infraction_code']==130)]

date_of_infraction  infraction_code        infraction_description  \
257773            20140527              130  STOP PROH TIME MUTUAL STREET   

        set_fine_amount  time_of_infraction location1            location2  \
257773                0                1153        RO  2184 EGLINTON AVE W   

        year  year2  
257773  2014      0

In [61]:
tpt_df[(tpt_df['infraction_code']==256)]

date_of_infraction  infraction_code          infraction_description  \
914850            20120427              256  LEAVE OVERNIGHT WITHOUT PERMIT   

        set_fine_amount  time_of_infraction location1    location2  year  \
914850                0                1246       NaN  3 MASSEY SQ  2012   

        year2  
914850      0

Looks like most of these $0 fine amt exceptions are pretty rare...

In [64]:
tpt_df_grouped[(tpt_df_grouped['location2']=='393 UNIVERSITY AVE') &(tpt_df_grouped['infraction_code']==15) ]

year  infraction_code           location2  count
206509   2008               15  393 UNIVERSITY AVE    465
592857   2009               15  393 UNIVERSITY AVE    285
962528   2010               15  393 UNIVERSITY AVE    318
1336055  2011               15  393 UNIVERSITY AVE    557
1706193  2012               15  393 UNIVERSITY AVE    710
1984529  2013               15  393 UNIVERSITY AVE    297
2257274  2014               15  393 UNIVERSITY AVE    478
2633081  2015               15  393 UNIVERSITY AVE    317

In [66]:
fine_amount_dict[15]

{'fine_amt': 100.0, 'fine_descp': 'PARK 3 M OF FIRE HYDRANT'}

Sweet lord this fire hydrant made $71,000 in 2012....maybe after this year they decided to paint the curb red??

In [71]:
tpt_df_grouped[(tpt_df_grouped['location2']=='361 UNIVERSITY AVE') &(tpt_df_grouped['infraction_code']==15)]

year  infraction_code           location2  count
206231   2008               15  361 UNIVERSITY AVE    246
592604   2009               15  361 UNIVERSITY AVE    173
962277   2010               15  361 UNIVERSITY AVE    188
1335785  2011               15  361 UNIVERSITY AVE    296
1705949  2012               15  361 UNIVERSITY AVE    373
1984403  2013               15  361 UNIVERSITY AVE    113
2257035  2014               15  361 UNIVERSITY AVE    276
2632861  2015               15  361 UNIVERSITY AVE    274

In [78]:
tpt_df.groupby(['year'])['set_fine_amount'].sum()#.reset_index()

year
2008    111153475
2009    105288625
2010    109898240
2011    114183140
2012    111449490
2013     42479400
2014    108971405
2015     99851365
Name: set_fine_amount, dtype: int64

In [79]:
tpt_df.groupby(['year','infraction_code'])['set_fine_amount'].sum()#.reset_index()

year  infraction_code
2008  1                    357090
      2                   1967850
      3                   9786870
      4                    362970
      5                  20747050
      6                   2689110
      7                    965500
      8                   6781080
      9                  10719600
      10                   221700
      11                     9045
      12                     9885
      13                     2310
      14                   187800
      15                  3905310
      16                   636440
      17                    12070
      18                     1070
      19                     1050
      20                      620
      21                     1230
      22                       40
      23                      740
      24                     1890
      25                   173460
      26                     8250
      28                   929340
      29                 13580130
      30                  

In [153]:
fine_amt_df=pd.DataFrame.from_dict(fine_amount_dict,orient='index')

fine_amt_df['infraction_code']=fine_amt_df.index #index is infraction code

fine_amt_df.sort_values(by='fine_amt',ascending=False)

fine_descp  fine_amt  infraction_code
358  STAND IN DISABLED NOT DROP OFF       450              358
350  PARK DISABLD LOADING NO PERMIT       450              350
368  PARK STR DISABLD LDG NO DROPOF       450              368
367  STAND ON-ST DISABL LDG NO PRMT       450              367
366  PARK ON-ST DISABLD LDG NO PRMT       450              366
365  STOP ON-STRT DISABLD NO PERMIT       450              365
364  STAND ON-STRT DISABL NO PERMIT       450              364
363  PARK ON-STRT DISABLD NO PERMIT       450              363
362  STAND DISABLD LDG NOT DROP OFF       450              362
361   PARK DISABLD LDG NOT DROP OFF       450              361
360  STAND DISABL LOADING NO PERMIT       450              360
359  PARK DISABLD LOADING NO PERMIT       450              359
349   PARK IN DISABLED NOT DROP OFF       450              349
357   PARK IN DISABLED NOT DROP OFF       450              357
356     STAND IN DISABLED NO PERMIT       450              356
355      PARK IN DISABLED NO PERMIT       450              355
354  PARK STR DISABLD LDG NO DROPOF       450              354
353  PARK ON-ST DISABLD LDG NO PRMT       450              353
352  PARK ON-STRT DISABLD NO PERMIT       450              352
369  STAND STR DISABL LDG NO DROPOF       450              369
351   PARK DISABLD LDG NOT DROP OFF       450              351
259   STOP/PARK DIS SPACE NO PERMIT       300              259
322    STOP NON-BUS - BUS LOAD ZONE       300              322
260   STOP/PARK DIS SPACE NO PERMIT       300              260
374       STAND BUS PROHIBITED TIME       300              374
373        STOP BUS PROHIBITED TIME       300              373
372        PARK BUS PROHIBITED TIME       300              372
334  PARK BUS PK SPC/MCHINE-NOFEEPD       300              334
347            PARK IN A FIRE ROUTE       250              347
220     PARK ON TRACK ON SNOW ROUTE       200              220
..                              ...       ...              ...
306  ANGLE PARK-TOO FAR FROM METER         30              306
305  PARK-F OF VEH TOO FAR FROM MTR        30              305
304   PARK-FRONT OF VEH NOT OPP MTR        30              304
303  PARK-R OF VEH TOO FAR FROM MTR        30              303
302   PAR PARK-R OF VEH NOT OPP MTR        30              302
301  PAR PARK - F TIRES NOT OPP MTR        30              301
318     PARK/LEAVE UNPD PARKING INV        30              318
13   PARK VEH TO WASH GREASE REPAIR        15               13
115     PARK METERED SPACE PARK LOT        15              115
56   FAIL TO PARK/STOP PARALLEL -LF        15               56
54   PARK/STOP OVER 30 CM FROM CURB        15               54
11   PARK/STOP OVER 30 CM FROM CURB        15               11
12            PARK VEHICLE-FOR SALE        15               12
52   FAIL TO ANGLE PARK/STOP DES SP        15               52
53    FAIL TO PARK/STOP PARALLEL TO        15               53
51   FAIL TO PARK/STOP AT 45 DG CRB        15               51
50    FAIL TO PARK/STOP NR RGHD HWY        15               50
49    FAIL TO PARK/STOP PARA RH HWY        15               49
48    FAIL TO PARK/STOP PAR TO CURB        15               48
2          PARK LONGER THAN 3 HOURS        15                2
55    FAIL TO PARK/STOP AS CLOSE AS        15               55
348      PARK IN DISABLED NO PERMIT         0              348
150      PARK VEH BEYOND DESIG AREA         0              150
258  STOP CONTRARY POSTED CONDITION         0              258
166           PARK ON TRINITY DRIVE         0              166
197  PARK NOT WITHIN PERMITTED TIME         0              197
130    STOP PROH TIME MUTUAL STREET         0              130
251    LEAVE IN PARK NOT DESIG AREA         0              251
252  PARK IN PARK - PROHIBITED AREA         0              252
256  LEAVE OVERNIGHT WITHOUT PERMIT         0              256

[246 rows x 3 columns]

In [100]:
tpt_grouped_df[(tpt_grouped_df['infraction_code']==15) & (tpt_grouped_df['year']==2012)] \
        .sort_values('fine_sum',ascending=False).head(n=10)

# tpt_grouped_df[tpt_grouped_df['year']==2012] \
#         .sort_values('fine_sum',ascending=False)

year  infraction_code           location2  count  fine_sum  \
1706193  2012               15  393 UNIVERSITY AVE    710     71000   
1707484  2012               15       5519 YONGE ST    445     44500   
1707541  2012               15    56 THE ESPLANADE    386     38600   
1705949  2012               15  361 UNIVERSITY AVE    373     37300   
1705602  2012               15     33 ELMHURST AVE    340     34000   
1701786  2012               15       112 MERTON ST    323     32300   
1709280  2012               15     99 ATLANTIC AVE    308     30800   
1709288  2012               15      99 ERSKINE AVE    287     28700   
1704511  2012               15     25 BELLEVUE AVE    272     27200   
1707224  2012               15       5150 YONGE ST    251     25100   

                      fine_descrp  
1706193  PARK 3 M OF FIRE HYDRANT  
1707484  PARK 3 M OF FIRE HYDRANT  
1707541  PARK 3 M OF FIRE HYDRANT  
1705949  PARK 3 M OF FIRE HYDRANT  
1705602  PARK 3 M OF FIRE HYDRANT  
1701786  PARK 3 M OF FIRE HYDRANT  
1709280  PARK 3 M OF FIRE HYDRANT  
1709288  PARK 3 M OF FIRE HYDRANT  
1704511  PARK 3 M OF FIRE HYDRANT  
1707224  PARK 3 M OF FIRE HYDRANT

In [104]:
mean_time_df=tpt_df.groupby(['year','infraction_code','location2'])['time_of_infraction'].mean().reset_index()

In [105]:
print (mean_time_df.shape)
print (tpt_grouped_df.shape)
tpt_merged_df=pd.merge(tpt_grouped_df, mean_time_df, on=['year','infraction_code','location2'], how='inner')
print (tpt_merged_df.shape)
tpt_merged_df.head()

(2833240, 4)
(2833240, 6)
(2833240, 7)


year  infraction_code           location2  count  fine_sum  \
0  2008                1    1  FOUR WINDS DR      1        30   
1  2008                1     1 9 MOBERLY AVE      1        30   
2  2008                1         1 BANGOR RD      2        60   
3  2008                1         1 BEAMAN RD      2        60   
4  2008                1  1 BEDFORD PARK AVE      8       240   

                      fine_descrp  time_of_infraction  
0  PARK FAIL TO DEPOSIT FEE METER            1033.000  
1  PARK FAIL TO DEPOSIT FEE METER            1106.000  
2  PARK FAIL TO DEPOSIT FEE METER            1640.000  
3  PARK FAIL TO DEPOSIT FEE METER            1291.000  
4  PARK FAIL TO DEPOSIT FEE METER            1428.625

In [129]:
tpt_merged_df=tpt_merged_df.rename(index=str, columns={"time_of_infraction": "mean_time_of_infraction" \
                                                               ,"location2":"street_address"})
tpt_merged_df.head()

year  infraction_code      street_address  count  fine_sum  \
0  2008                1    1  FOUR WINDS DR      1        30   
1  2008                1     1 9 MOBERLY AVE      1        30   
2  2008                1         1 BANGOR RD      2        60   
3  2008                1         1 BEAMAN RD      2        60   
4  2008                1  1 BEDFORD PARK AVE      8       240   

                      fine_descrp  mean_time_of_infraction  
0  PARK FAIL TO DEPOSIT FEE METER                 1033.000  
1  PARK FAIL TO DEPOSIT FEE METER                 1106.000  
2  PARK FAIL TO DEPOSIT FEE METER                 1640.000  
3  PARK FAIL TO DEPOSIT FEE METER                 1291.000  
4  PARK FAIL TO DEPOSIT FEE METER                 1428.625

In [148]:
##create a DF with the top 500 grossing fire hydrants each year...

final_df=None
df_list=[]
for year in range(2008,2016):
    df_list.append(tpt_merged_df[(tpt_merged_df['infraction_code']==15) & (tpt_merged_df['year']==year)] \
            .sort_values('fine_sum',ascending=False).head(n=500))

final_df= pd.concat(df_list, axis=0)


##time formatting is weird; 5 --> 12:05am, 30 --> 12:30am, 15:20 --> 3:20pm etc.
def round_time(df):
    time_str=str(df['mean_time_of_infraction'])
    time_len=len(time_str.split('.')[0])
    if time_len==1:
        rnded_time=0
    elif time_len==2:
        if float(time_str)<30:
            rnded_time=0
        else:
            rnded_time=1
    elif time_len==3:
        rnded_time=round(float(time_str[0:1]+'.'+time_str[1:3]),0)
    elif time_len==4:
        rnded_time=round(float(time_str[0:2]+'.'+time_str[2:4]),0)
    else:
        rnded_time=0
    return rnded_time


final_df['mean_hour']=final_df.apply(round_time,axis=1)
final_df=final_df.drop(['mean_time_of_infraction'],axis=1)

final_df

year  infraction_code        street_address  count  fine_sum  \
206968   2008               15             43 ELM ST    869     86900   
207724   2008               15         5100 YONGE ST    496     49600   
206509   2008               15    393 UNIVERSITY AVE    465     46500   
205795   2008               15       33 ELMHURST AVE    397     39700   
208114   2008               15      56 THE ESPLANADE    384     38400   
206668   2008               15  40 ORCHARD VIEW BLVD    327     32700   
201055   2008               15         112 MERTON ST    300     30000   
207742   2008               15         5150 YONGE ST    299     29900   
208370   2008               15   6 SPRING GARDEN AVE    262     26200   
204778   2008               15    26 WELLINGTON ST E    248     24800   
206231   2008               15    361 UNIVERSITY AVE    246     24600   
208042   2008               15         5519 YONGE ST    193     19300   
203011   2008               15       184 SPADINA AVE    174     17400   
210614   2008               15           VILLIERS ST    168     16800   
205513   2008               15             31 ELM ST    158     15800   
210145   2008               15      98 THE ESPLANADE    155     15500   
202495   2008               15         160 MUTUAL ST    150     15000   
204407   2008               15         2409 YONGE ST    147     14700   
209853   2008               15           9 CHURCH ST    144     14400   
206855   2008               15        417 SPADINA RD    144     14400   
205612   2008               15        318 QUEEN ST W    142     14200   
207761   2008               15   518 PALMERSTON BLVD    140     14000   
210237   2008               15            BOROUGH DR    138     13800   
202737   2008               15         171 KING ST E    136     13600   
205360   2008               15       30 GLENDALE AVE    133     13300   
206212   2008               15    36 WELLINGTON ST E    132     13200   
204505   2008               15       25 BELLEVUE AVE    130     13000   
204109   2008               15       227 VICTORIA ST    127     12700   
201440   2008               15     123 SUNNYSIDE AVE    125     12500   
207670   2008               15         5050 YONGE ST    125     12500   
...       ...              ...                   ...    ...       ...   
2628348  2015               15            1 SCOTT ST     12      1200   
2634934  2015               15      68 NORTHCOTE AVE     12      1200   
2636079  2015               15       EAST LIBERTY ST     12      1200   
2633025  2015               15  386 RONCESVALLES AVE     12      1200   
2628617  2015               15         106 WALTON ST     12      1200   
2634971  2015               15           69 LAING ST     12      1200   
2629463  2015               15        14 SAYWELL AVE     12      1200   
2631291  2015               15       238 BERKELEY ST     12      1200   
2634819  2015               15     650 PARLIAMENT ST     12      1200   
2634374  2015               15      56 TEMPERANCE ST     12      1200   
2630893  2015               15       2120 QUEEN ST E     12      1200   
2629922  2015               15    160 DAN LECKIE WAY     12      1200   
2633700  2015               15         47 DAVIES AVE     12      1200   
2634222  2015               15      540 SPADINA CRES     12      1200   
2632002  2015               15         29 NIAGARA ST     12      1200   
2630843  2015               15         21 MELINDA ST     12      1200   
2635330  2015               15      79 WINCHESTER ST     12      1200   
2634391  2015               15        561 QUEEN ST W     12      1200   
2634866  2015               15       660 CRAWFORD ST     12      1200   
2632629  2015               15         3397 YONGE ST     12      1200   
2635798  2015               15  94 EVELYN WIGGINS DR     12      1200   
2636364  2015               15          ST JOSEPH ST     12      1200   
2634638  2015               15       601 KINGSTON RD     12      12

In [158]:
locations_1=final_df['street_address'].unique()
len(locations_1)

1266

In [150]:
##create a DF with the top 500 grossing spots each year...
final_df_2=None
df_list=[]
for year in range(2008,2016):
    df_list.append(tpt_merged_df[tpt_merged_df['year']==year].sort_values('fine_sum',ascending=False).head(n=500))

final_df_2= pd.concat(df_list, axis=0)

##time formatting is weird; 5 --> 12:05am, 30 --> 12:30am, 15:20 --> 3:20pm etc.
def round_time(df):
    time_str=str(df['mean_time_of_infraction'])
    time_len=len(time_str.split('.')[0])
    if time_len==1:
        rnded_time=0
    elif time_len==2:
        if float(time_str)<30:
            rnded_time=0
        else:
            rnded_time=1
    elif time_len==3:
        rnded_time=round(float(time_str[0:1]+'.'+time_str[1:3]),0)
    elif time_len==4:
        rnded_time=round(float(time_str[0:2]+'.'+time_str[2:4]),0)
    else:
        rnded_time=0
    return rnded_time


final_df_2['mean_hour']=final_df_2.apply(round_time,axis=1)
final_df_2=final_df_2.drop(['mean_time_of_infraction'],axis=1)

final_df_2


year  infraction_code          street_address  count  fine_sum  \
53976    2008                3        25 THE WEST MALL   8291    248730   
50294    2008                3        1750 FINCH AVE E   7525    225750   
90392    2008                5            20 EDWARD ST   4328    173120   
52076    2008                3        2075 BAYVIEW AVE   5394    161820   
189893   2008                9            60 MURRAY ST   2683    160980   
401943   2008              347         45 OVERLEA BLVD    479    119750   
52075    2008                3         2075 BAYVIEW AV   3916    117480   
169512   2008                8         244 VICTORIA ST   1820    109200   
96905    2008                5           2401 YONGE ST   2552    102080   
194386   2008                9                JAMES ST   1664     99840   
181384   2008                9        200 ELIZABETH ST   1542     92520   
190618   2008                9      66 WELLINGTON ST W   1510     90600   
401342   2008              347          31 TAPSCOTT RD    361     90250   
401487   2008              347        3401 DUFFERIN ST    355     88750   
401252   2008              347          300 BOROUGH DR    351     87750   
402185   2008              347           5100 YONGE ST    349     87250   
206968   2008               15               43 ELM ST    869     86900   
402836   2008              347         85 ELLESMERE RD    346     86500   
400314   2008              347     1921 EGLINTON AVE E    335     83750   
168739   2008                8           200 KING ST E   1344     80640   
167484   2008                8          15 ASQUITH AVE   1280     76800   
133887   2008                5          74 VICTORIA ST   1896     75840   
81905    2008                5         158 VICTORIA ST   1867     74680   
400851   2008              347        25 THE WEST MALL    298     74500   
171430   2008                8    40 ORCHARD VIEW BLVD   1208     72480   
47976    2008                3        1265 MILITARY TR   2409     72270   
168680   2008                8       20 EGLINTON AVE W   1200     72000   
381887   2008              250  1675 LAKE SHORE BLVD E    679     71295   
173489   2008                8             78 HOLLY ST   1167     70020   
402043   2008              347           4968 YONGE ST    276     69000   
...       ...              ...                     ...    ...       ...   
2791960  2015              403           21 BALMUTO ST    106     15900   
2795889  2015              403           5100 YONGE ST    106     15900   
2577273  2015                5        51 WINCHESTER ST    396     15840   
2616499  2015                9            11 NELSON ST    264     15840   
2500210  2015                2   34 LITTLE NORWAY CRES   1055     15825   
2782733  2015              355           15 PROVOST DR     35     15750   
2782069  2015              347           6464 YONGE ST     63     15750   
2620104  2015                9         3 PARK HOME AVE    262     15720   
2617692  2015                9        150 ELIZABETH ST    262     15720   
2585237  2015                5          70 CHESTNUT PK    392     15680   
2593640  2015                5             COXWELL AVE    391     15640   
2557065  2015                5          23 VICTORIA ST    391     15640   
2832538  2015              415        50 CUMBERLAND ST    390     15600   
2796780  2015              403           590 KING ST W    104     15600   
2795504  2015              403          483 QUEEN ST W    104     15600   
2787684  2015              403          1005 KING ST W    104     15600   
2787951  2015              403       105 ADELAIDE ST W    104     15600   
2762209  2015              210         40 GERRARD ST E    518     15540   
2593561  2015                5               CHURCH ST    388     15520   
2780201  2015              347          151 TORYORK DR     62     15500   
2611886  2015                8               2 ORDE ST    258     15480   
2521420  2015                3    

In [159]:
locations_2=final_df_2['street_address'].unique()
len(locations_2)

1141

This gives us 2400 unique locations (under google api limit for one day!) for two different maps.

1st map will be highest grossing fire hydrants
- Visual Encodings -size --> fine amount -location--> location -colour -->infraction time
2nd map will be highest grossing spots in general
- Visual Encodings -size --> fine amount -location--> location -colour --> infraction type

In [160]:
req_url='https://maps.googleapis.com/maps/api/geocode/json?address={0}&key={1}'
api_key = Ugen.ConfigSectionMap('googlemaps')['key']

def get_coords(addr):
    address=addr +' Toronto, Ontario, Canada'
    try:
        response = requests.get(req_url.format(address,api_key))
        response_dict = response.json()
        coords=response_dict['results'][0]['geometry']['location']
        if response_dict['status'] == 'OK':
#             print (response_dict)
            return coords['lat'],coords['lng'],response_dict['results'][0]['formatted_address'].replace(',','')
    except Exception as err:
        print (str(err))
        pass
    return None,None,None

####TO DO
#need to change this so it gets all the different results..
#then will need to update dataframes using these dicts...

def dummy(x):
    return x*1,x*2,x*3


locations_1_dict={addr:get_coords(addr) for addr in locations_1}
locations_2_dict={addr:get_coords(addr) for addr in locations_2}



{5: (5, 10, 15),
 6: (6, 12, 18),
 7: (7, 14, 21),
 8: (8, 16, 24),
 9: (9, 18, 27)}

In [ ]:
##update final_df's with google data

final_df['lat']=locations_1_dict['street_address'][0]
final_df['lng']=locations_1_dict['street_address'][1]
final_df['full_address']=locations_1_dict['street_address'][2]


final_df_2['lat']=locations_2_dict['street_address'][0]
final_df_2['lng']=locations_2_dict['street_address'][1]
final_df_2['full_address']=locations_2_dict['street_address'][2]

In [154]:
# final_df.to_csv('./top_hydrants.csv')
# final_df_2.to_csv('./top_spots.csv')

# fine_amt_df.to_csv('./fine_amount_coding.csv')

General notes:
- apparently hydrant fines changed from 30to100 part way through 2008
- with your current method your multipying all fines by $100 so numbers will be a bit off...


To Do:
- interactive dashboard
- time series graphs based on user selection of location and infraction code (or show multiple infraction codes per graph)
- overall summary (fine sums per year, fine sums per year per infrac_code)

Interactive Map:
- Visual Encodings
     -size --> fine amount
     -location--> location
     -colour --> infraction code OR infraction time
     -pop-up menu --> count, fine_amt & fine description